src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -3931,20 +3932,30 @@ CreateBasicTask(uint64 jobId, uint32 taskId, TaskType taskType, char *queryStrin
  * UpdateRangeTableAlias walks over each fragment in the given fragment list,
  * and creates an alias that represents the fragment name to be used in the
  * query. The function then updates the corresponding range table entry with
- * this alias.
+ * this alias along with rewriting other references in the Query that reference
+ * the relation (ie, ::regclass casts) to include the shardId.
  */
 static void
-UpdateRangeTableAlias(List *rangeTableList, List *fragmentList)
+UpdateRangeTableAlias(List *rangeTableList, List *fragmentList, Query *query)
 {
 	ListCell *fragmentCell = NULL;
 	foreach(fragmentCell, fragmentList)
 	{
 		RangeTableFragment *fragment = (RangeTableFragment *) lfirst(fragmentCell);
 		Index rangeTableId = fragment->rangeTableId;
 		RangeTblEntry *rangeTableEntry = rt_fetch(rangeTableId, rangeTableList);
+		uint64 shardId;
+		char *relationName;
 
-		Alias *fragmentAlias = FragmentAlias(rangeTableEntry, fragment);
+		Alias *fragmentAlias = FragmentAlias(rangeTableEntry, fragment, &relationName,
+											 &shardId);
 		rangeTableEntry->alias = fragmentAlias;
+
+		if (relationName != NULL)
+		{
+			Assert(shardId != 0);
+			AppendShardIdToRelationReferences(relationName, shardId, (void **) &query);
```

Comment:
Why do we need this in readonly queries? Yet another iteration over the whole tree is far from free.